components/TextFields/src/MDCTextField.m

```diff
@@ -0,0 +1,439 @@
+  // Y that works.
+  CGFloat actualY =
+      (CGRectGetHeight(bounds) / 2.f) - MDCRint(MAX(self.font.lineHeight,
+                                                    self.placeholderLabel.font.lineHeight) /
+                                                2.f);  // Text field or placeholder
+  actualY = textContainerInset.top - actualY;
+  textRect.origin.y = actualY;
+
+  if (self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    // Now that the text field is laid out as if it were LTR, we can flip it if necessary.
+    textRect = MDCRectFlippedForRTL(textRect, CGRectGetWidth(bounds),
+                                    UIUserInterfaceLayoutDirectionRightToLeft);
+  }
+
+  return textRect;
+}
+
+- (CGRect)editingRectForBounds:(CGRect)bounds {
+  // First the textRect is loaded. Then it's shaved for cursor and/or clear button.
+  CGRect editingRect = [self textRectForBounds:bounds];
+
+  // The textRect comes to us flipped for RTL (if RTL) so we flip it back before adjusting.
+  if (self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    editingRect = MDCRectFlippedForRTL(editingRect, CGRectGetWidth(bounds),
+                                       UIUserInterfaceLayoutDirectionRightToLeft);
+  }
+
+  // UITextFields show EITHER the clear button or the rightView. If the rightView has a superview,
+  // then it's being shown and the clear button isn't.
+  if (self.rightView.superview) {
+    editingRect.size.width += MDCTextInputEditingRectRightViewPaddingCorrection;
+  } else {
+    if (self.text.length > 0) {
+      CGFloat clearButtonWidth = CGRectGetWidth(self.clearButton.bounds);
+
+      // The width is adjusted by the padding twice: once for the right side, once for left.
+      clearButtonWidth += 2 * MDCTextInputClearButtonImageBuiltInPadding;
+
+      // The clear button's width is already subtracted from the textRect.width if .always or
+      // .unlessEditing.
+      switch (self.clearButtonMode) {
+        case UITextFieldViewModeUnlessEditing:
+          editingRect.size.width += clearButtonWidth;
+          break;
+        case UITextFieldViewModeWhileEditing:
+          editingRect.size.width -= clearButtonWidth;
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  if (self.mdc_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+    editingRect = MDCRectFlippedForRTL(editingRect, CGRectGetWidth(bounds),
+                                       UIUserInterfaceLayoutDirectionRightToLeft);
+  }
+
+  if ([self.fundament.positioningDelegate
+          respondsToSelector:@selector(editingRectForBounds:defaultRect:)]) {
+    editingRect =
+        [self.fundament.positioningDelegate editingRectForBounds:bounds defaultRect:editingRect];
+  }
+
+  return editingRect;
+}
+
+- (CGRect)clearButtonRectForBounds:(CGRect)bounds {
+  return self.clearButton.frame;
```

Comment:
I've never run into that problem. But I also have no proof that UIKit does it on its own. I'm happy to add it.